SQL: Add macro support in select case

zoltanbedi opened the pull request:
**This is behind a feature flag: `sqlQuerybuilderFunctionParameters`**


**What is this feature?**

This PR updates the Visual Query Builder for SQL data sources (MySQL, PostgreSQL, MSSQL, Influx Flight SQL) to add the possibility to add parameters to functions in the select clause. This is essential to be able to add a `$__timeGroup`,`$__timeGroupAlias` macros.

Changed the order for the column selection so the user would first select the data operation and then the column. It felt more intuitive like that.


<img width="1189" alt="screenshot-sql-builder-time-series-query" src="https://github.com/grafana/grafana/assets/13729989/be826efe-ddac-4f55-a025-0feb88db49c9">



### TODO: 
- [x] Update documentation

**Which issue(s) does this PR fix?**:

Fixes #83593

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

zoltanbedi committed:
Feat: timeGroup macro handling in VQB

zoltanbedi committed:
Add tests

zoltanbedi committed:
Add functions to SQL ds

zoltanbedi committed:
Fix lint errors

gabor reviewed (commented):
very nice 👍 

some notes:
1. is this approach something we could also do later in `bigquery` and in https://github.com/grafana/plugin-ui/blob/main/src/components/QueryEditor/visual-query-builder/SQLSelectRow.tsx ? i assume it is so, i'm just doublechecking, it would be great to choose an approach that we can later add to the other places too.
2. i wonder if we should slightly modify the `function()` thing, so that for the functions it could also say how many params it has or something, in other words, would be great, if, when choosing `$__timeGroup` for example, it would already add a second param, and default it to `$__interval`. of course, we can improve this incrementally 👍 
3. related to [2]. question is, do we merge this directly into main-branch, and then improve it continuously, or we somehow "hide" it until we add more polish to it? (we could put this SOMEHOW behind an experimental feature-toggle for example) WDYT?

zoltanbedi committed:
Merge remote-tracking branch 'origin/main' into zoltan/time-group-vqb

zoltanbedi committed:
Add feature toggle

diegommm commented:
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

zoltanbedi committed:
Add rendering based on object

zoltanbedi committed:
Merge remote-tracking branch 'origin/main' into zoltan/time-group-vqb

zoltanbedi committed:
Fix lint

zoltanbedi committed:
Fix CI failures

zoltanbedi commented on packages/grafana-sql/src/components/visual-query-builder/SQLSelectRow.tsx:
```diff
@@ -18,7 +18,7 @@ interface SQLSelectRowProps {
 export function SQLSelectRow({ fields, query, onQueryChange, db }: SQLSelectRowProps) {
   const { onSqlChange } = useSqlChange({ query, onQueryChange, db });
-  const functions = [...COMMON_AGGREGATE_FNS, ...(db.functions?.() || [])].map(toOption);
+  const functions = [...COMMON_FNS, ...(db.functions?.() || [])].sort().map(toOption);
```
I moved this code elsewhere. So now all datasources import these COMMON_FNS and then the db.functions are used instead.

zoltanbedi commented:
> very nice 👍
> 
> some notes:
> 
> 1. is this approach something we could also do later in `bigquery` and in https://github.com/grafana/plugin-ui/blob/main/src/components/QueryEditor/visual-query-builder/SQLSelectRow.tsx ? i assume it is so, i'm just doublechecking, it would be great to choose an approach that we can later add to the other places too.
> 2. i wonder if we should slightly modify the `function()` thing, so that for the functions it could also say how many params it has or something, in other words, would be great, if, when choosing `$__timeGroup` for example, it would already add a second param, and default it to `$__interval`. of course, we can improve this incrementally 👍
> 3. related to [2]. question is, do we merge this directly into main-branch, and then improve it continuously, or we somehow "hide" it until we add more polish to it? (we could put this SOMEHOW behind an experimental feature-toggle for example) WDYT?

@gabor thanks for the feedback. I moved all these changes behind a `sqlQuerybuilderFunctionParameters` feature toggle.

I changed the structure a bit. So now the columns are rendered with the following logic. If the selected function or macro is from the list we will have the parameters so we render those if it is not we render column and the user can add as many (should we limit this?) parameters as they wish.

This still needs some more tests and documentation changes as well. Please let me know what you think.

zoltanbedi committed:
Fix tests

zoltanbedi committed:
Merge remote-tracking branch 'origin/main' into zoltan/time-group-vqb

ivanahuckova commented on packages/grafana-sql/src/components/visual-query-builder/SelectCustomFunctionParameters.tsx:
```diff
@@ -0,0 +1,137 @@
+      {renderParameters(currentColumnIndex)}
+      <Button
+        type="button"
+        onClick={addParameter(currentColumnIndex)}
```
personal nit, I got here a little bit confused why is `addParameter` and also `removeParameter` invoked, but then I noticed that they both are:
```
  const removeParameter = useCallback(
    (columnIndex: number, index: number) => () => {
```

Any reason why not just 
```
  const removeParameter = useCallback(
    (columnIndex: number, index: number) => {

		onClick={() => addParameter(currentColumnIndex)}
```

Feel free to ignore, I was literally just curious.

zoltanbedi replied:
I think this just a matter of taste how you write this. I can change it so it won't be confusing.

ivanahuckova replied:
No need to change, I was just curious!

ivanahuckova reviewed (commented):
I played with the editor and reviewed a code. I have couple comments  and questions and notes:

- I really like how you changed that in aggregations you write query in the similar way as you read it. 
- I noticed that the editor does not wrap correctly. This is an issue also with current editor, but because that one is smaller (always has just 2 inputs, it is less noticeable)
  
<img width="848" alt="image" src="https://github.com/grafana/grafana/assets/30407135/b1c664cf-ddad-4f72-b136-a285828f257c">

- We are mixing aggregations and macros and putting new macros on top of the list. Are we sure that macros are more used than aggregations are and that most of the users know what they mean? Because I personally understand `AVG`, `COUNT`, ... , but I don't know what `$__timeGroupAlias` means. 
- What about `Aggregation / Macro - optional` -> `Data operations - optional`? 

The code looks good! ✨ But I'd leave to developers who have more experience with this codebase and can review more thoroughly to ✅ .

zoltanbedi committed:
Merge remote-tracking branch 'origin/main' into zoltan/time-group-vqb

zoltanbedi commented:
> I noticed that the editor does not wrap correctly. This is an issue also with current editor, but because that one is smaller (always has just 2 inputs, it is less noticeable)

You are right, I haven't noticed this. This is actually a bug in grafana experimental. I'm going to open an issue there.

zoltanbedi committed:
Address review comments

zoltanbedi committed:
Add docs

zoltanbedi committed:
Merge remote-tracking branch 'origin/main' into zoltan/time-group-vqb

zoltanbedi committed:
Fix JSX runtime warnings

zoltanbedi committed:
Remove docs part that mentions suggest more macros

jdbaldry commented on docs/sources/shared/datasources/sql-query-builder-macros.md:
```diff
@@ -0,0 +1,19 @@
+    - oss
+---
+
+#### Macros
```
Note if you need to include this in places where it may not be an H4, you can use the `leveloffset` parameter to offset the heading: https://grafana.com/docs/writers-toolkit/write/shortcodes/#docsshared. Only positive offsets are implemented at the moment so it may make sense to start this as an H1. Or you can ask me to do the work to implement the negative offsets :D

zoltanbedi replied:
Good to know, thanks Jack.

jdbaldry commented on docs/sources/shared/datasources/sql-query-builder-macros.md:
```diff
@@ -0,0 +1,19 @@
+
+#### Macros
+
+You can enable macros support in the select clause. This allows you to create timeseries queries.
+
+{{< docs/experimental product="Macros support in visual query builder" featureFlag="`sqlQuerybuilderFunctionParameters`" >}}
+
+Use the **Data operations** dropdown to select a macro like `$__timeGroup` or `$__timeGroupAlias`. Select a time column from the **Column** dropdown and a time interval from the **Interval** dropdown to create a time series query.
+
+{{< figure src="/media/docs/grafana/data-sources/screenshot-sql-builder-time-series-query.png" class="docs-image--no-shadow" caption="SQL query builder time series query" >}}
+
+You can also add custom value to the **Data operations**. For example a function that is not in the dropdown list. This will allow you to add any number of parameters.
```
- I think these are all adjectival forms of "time series" so we should use "time-series": https://grafana.com/docs/writers-toolkit/write/style-guide/word-list/#time-series
- It's also "drop-down" rather than "dropdown": https://grafana.com/docs/writers-toolkit/write/style-guide/word-list/#drop-down
- I've also kept in the present tense (avoiding "will:) and made some minor edits for readability

```suggestion
You can enable macros support in the select clause to create time-series queries.

{{< docs/experimental product="Macros support in visual query builder" featureFlag="`sqlQuerybuilderFunctionParameters`" >}}

Use the **Data operations** drop-down to select a macro like `$__timeGroup` or `$__timeGroupAlias`.
Select a time column from the **Column** drop-down and a time interval from the **Interval** drop-down to create a time-series query.

{{< figure src="/media/docs/grafana/data-sources/screenshot-sql-builder-time-series-query.png" class="docs-image--no-shadow" caption="SQL query builder time-series query" >}}

You can also add custom value to the **Data operations**.
For example, a function that's not in the drop-down list.
This allows you to add any number of parameters.
```

zoltanbedi committed:
Update docs/sources/shared/datasources/sql-query-builder-macros.md

Co-authored-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry reviewed (approved):
Nice, LGTM

zoltanbedi committed:
Add smoke test for this feature

zoltanbedi committed:
Merge remote-tracking branch 'origin/main' into zoltan/time-group-vqb

zoltanbedi committed:
lint

zoltanbedi committed:
Add supported macros to influx

ivanahuckova reviewed (approved):
LGTM from me! Btw this looks :chefs-kiss:

<img width="297" alt="image" src="https://github.com/grafana/grafana/assets/30407135/950e8efe-36d3-4ca5-b5e8-e7fe38669154">

ivanahuckova commented:
Let's wait with merging this until @zoltanbedi is back. We talked internally and we think that the UI should not change (the flipped order of aggregation and column) if the toggle is not turned on.

zoltanbedi commented:
I'm back from parental leave and going to wrap up this asap.

zoltanbedi committed:
Merge remote-tracking branch 'origin/main' into zoltan/time-group-vqb

zoltanbedi committed:
Merge remote-tracking branch 'origin/main' into zoltan/time-group-vqb

aangelisc reviewed (approved):
Awesome work @zoltanbedi. I've tested this with both MSSQL and Influx - LGTM for MSSQL! Unfortunately it's exposed a bug in Influx but that isn't as a result of this PR so I think you're good to go 😊

Sihal3 commented:
Awesome -- was just looking at why simple time series queries were so easy to make using InfluxDB and not for Timescale. Went down a rabbit hole and found than an excellent query builder had somehow been dropped in 9.2. Glad to see it coming back.

zoltanbedi committed:
Merge remote-tracking branch 'origin/main' into zoltan/time-group-vqb

zoltanbedi committed:
Add setupTests.ts to include in tsconfig.json

zoltanbedi committed:
Import jest-dom instead of setupTests.ts
